c2rust-analyze/src/c_void_casts.rs

```diff
@@ -403,59 +426,99 @@ impl<'tcx> CVoidCasts<'tcx> {
-                    ),
-                    // For [`CVoidCastDirection::To`], we only count
-                    // a cast to `*c_void` from an arbitrary type in the same block,
-                    // searching backwards.
-                    To => Self::find_last_cast(&bb_data.statements, c_void_ptr),
-                };
-                if let Some((statement_index, cast)) = cast {
+                } = &term.kind
+                {
+                    let func_ty = func.ty(&body.local_decls, tcx);
+                    let directions = CVoidCastDirection::from_callee(ty_callee(tcx, func_ty));
+
+                    for direction in directions.iter().copied() {
+                        use CVoidCastDirection::*;
+
+                        match direction {
+                            From => {
+                                let c_void_ptr =
+                                    CVoidPtr::checked(*destination, &body.local_decls, tcx);
+                                if let Some((statement_index, cast)) = Self::find_first_cast(
+                                    &body.basic_blocks()[target.unwrap()].statements,
+                                    c_void_ptr,
+                                ) {
+                                    self.insert_cast(
+                                        direction,
+                                        Location {
+                                            statement_index,
+                                            block: target.unwrap(),
+                                        },
+                                    );
+                                    self.insert_call(
+                                        direction,
+                                        terminator_location(block, bb_data),
+                                        cast,
+                                    );
+                                }
+                            }
+                            To => {
+                                for arg in args {
+                                    if let Some(place) = arg.place() {
+                                        if let Some(c_void_ptr) = CVoidPtr::checked_optional(
+                                            place,
+                                            &body.local_decls,
+                                            tcx,
+                                        ) {
+                                            self.find_and_insert_pred_cast(
+                                                body, c_void_ptr, direction, block, bb_data,
+                                            );
+                                        }
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /// Performs BFS in the CFG to find where the c_void_ptr
+    /// was cast to that type
+    fn find_and_insert_pred_cast(
```

Comment:
I'm not a fan of using BFS for this - it seems error prone.  Is it the case that the `CVoidPtr` `Place` is always a temp local, which is written only once?  I think it should be for code like `memset(x as *mut c_void, ...)`.  If there's only one assignment, then we don't need to worry about where exactly the cast is located relative to the call, and instead we can just scan the whole body to find it.